### DISCUSSION TITLE
Passing a body_image reference in body_params

### DISCUSSION THREAD

**Discussion by rfow** [Miksus/red-mail#44] [Q&A] [answered]
I'd like to combine a number of features in Red Mail to create an email report with a series of images by looping through body_params containing image references.

I'm facing the obstacle that the image name in body_params has to be referenced as a string which doesnt work when used in the html template itself.

If anyone has an idea on how this might be overcome, I'd really appreciate any pointers.

Thanks vm,
Richard

    email.send(
        subject='this is a test!',
        sender='me@example.com',
        receivers=['you@example.com'],
        html="""
            <h1>Chart overview</h1>
            <p>See below charts</p>
            {% for c in charts %}
                {{ c[chart_title] }}
                {{ c[chart_ref] }}
            {% endfor %}
        """,
        body_params={
            'charts': [
                {'chart_title': 'First Chart',
                 'chart_ref': 'image1'},
                {'chart_title': 'Second Chart',
                 'chart_ref': 'image2'}
            ]
        },
        body_images={
            'image1': 'path/to/image1.png',
            'image2': 'path/to/image2.png'
        }
    )

**Comment by Miksus** [comment, marked as answer]
This seems an interesting dilemma that I haven't thought of. What is done with the body images (if I remember correctly on top of my head) is that first, CIDs are defined for each of them (Content-IDs) using a function from the email library, they are wrapped with utility class BodyImage that is actually passed to the HTML template using Jinja as "direct" arguments and then the actual image is read/parsed and put as an inline image to the email body.

For time being, unfortunately, I cannot think of a convenient solution. 

I think we could do something about this though. Probably passing the images as an argument called just ``body_images`` if one wishes to loop through them, and why not do the same with body_params as well. Or maybe the BodyImage utility class could be more exposed and used more in the process so that it could be passed directly to body_params and then the sender would include those automatically as inline images but not sure how that could be done at the moment.

**Reply by rfow** [reply]
Thanks vm for your reply (and the library itself which makes the process considerably more straightforward!).

A fairly brutish workaround I used in the meantime (in a forked repo) was to adjust the CID naming conventions to something that wasnt random but also wasnt too simple.

That mighn't be the best solution overall - I'm not really familiar with how inline images are handled at the server level etc and what edge cases might arise, but it did mean that I could calculate what the CID would be and use that directly in the template.

Thanks again for your help!
